File: Amazon_LeetCode_Python_C++/LinkedList/2_add_two_numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {

private:
    ListNode* additions(ListNode* l1, ListNode* l2, int carry) {
        if (!l1 && !l2 && !carry) {
            return nullptr;
        }
        int sum = carry + (l1 ? l1->val : 0) + (l2 ? l2->val : 0);
        carry = sum / 10;
        sum %= 10;
        ListNode *p = new ListNode(sum);
        p->next = additions((l1 ? l1->next : nullptr), (l2 ? l2->next : nullptr), carry);
        return p;
    }
    
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        return additions(l1, l2, 0);
    }
};

File: LeetCode/Explore/TopInterview/Amazon/LinkedList/2_add_two_numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
  
    def additions(self, l1: Optional[ListNode], l2: Optional[ListNode], carry: int):
      if l1 == None and l2 == None and carry == 0:
        return None
      
      sum = carry
      
      if l1:
        sum += l1.val
        
      if l2:
        sum += l2.val
        
      carry = sum // 10
      sum = sum % 10
      
      p = ListNode(sum)
      
      if l1 and l1.next and l2 and l2.next:
        p.next = self.additions(l1.next, l2.next, carry)
      elif not (l1 and l1.next) and l2 and l2.next:
        p.next = self.additions(None, l2.next, carry)
      elif l1 and l1.next and not (l2 and l2.next):
        p.next = self.additions(l1.next, None, carry)
      elif not (l1 and l1.next) and not (l2 and l2.next):
        p.next = self.additions(None, None, carry)
      
      return p
  
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        return self.additions(l1, l2, 0)
